proposals/gateway-plusplus-phase1.md

```diff
@@ -0,0 +1,122 @@
+
+- Document configuring IPFS Gateway to pass Pinning API requests to a cluster or other IPFS node with the Pinning API enabled.
+- Configure and deploy our own gateway with this configuration. (We won’t be handing out auth tokens to anyone who doesn’t already have one, this is just to eat our own dogfood).
+
+#### What does done look like?
+
+Each work item in Phase 1 is something nft.storage needs so we have a builtin "first user" who would have all their requirements satisfied by this proposal.
+
+Once this is deployed you'll be able to write thin client interfaces with only a gateway URL and bearer token as configuration.
+
+####  What does success look like?
+
+Beyond just having satisfied the needs of nft.storage, this should also satisfy the needs that Cloudflare has and we would hopefully see support for IPFS in Cloudflare Workers as a thin
+client to their Gateway.
+
+#### Alternatives
+
+There are alternative approaches to building thin clients. The proposals around changing/improving the RPC API could be designed for this purpose,
+but the RPC interface is a larger API (not very "thin") and wasn't designed for multi-tenant. There would probably be **more** work involved in getting
+the RPC interface to support these users than this proposal.
+
+#### Counterpoints &amp; pre-mortem
+
+While this maps well to where web developers are today, it's not a "pure p2p" approach to solving problems. We're beefing up the ability to rely on large IPFS nodes that end up
+being federated rather than fully decentralized.
+
+#### Dependencies/prerequisites
+
+None
+
+#### Impact
+
+*Just* for nft.storage, 7. Overall, a 9 or 10, this should unlock numerous thin clients to be built and will allow developers to leverage a lot more of their existing infrastructure.
+
+#### Confidence
+
+Confidence is a 10 for "nft.storage needs this" but is an 8 or 9 for "this is absolutely the right thin client interface."
+
+## Future opportunities
+### New Thin Clients
+
+Rather than thinking of a thin client as the current IPFS API with less code, let’s consider what a smaller overall API profile looks like for this new interface.
+
+The JS library work we’ve been doing for the last year could be leveraged for a much smaller and higher impact JS library built against the Gateway++ interface.
+
+Since all writes that pass through the CAR file interface are transactional we now have availability guarantees for all of the writes.
+
+We can do file encoding in the client (we’re going to have to write this anyway for [nft.storage](http://nft.storage) in order to get over the 100MB Cloudflare Worker limit) using the latest codecs.
+
+Finally, since the protocol is so simple we should expect to see serveral thin clients designed to meet a variety of user needs and integrations.
+
+### Content Routing for Large Providers
+
+Gateways and large providers need to be directly peered since large providers have too much content to provide in the DHT.
```

Comment:
It’ll help, but we’re throwing incremental improvements as an exponential problem. nft.storage will have too many CIDs to keep in the DHT by the end of the month even with roots only and the improvements Adin made that havent been released.

We’ve come to the same conclusion other large providers like Pinata came to, we can’t support the DHT with this much content.

But this is going to work out because content discovery has always been about more than just the DHT. We should work on a protocol for a federation of large providers to use **and** continue to improve the DHT for a larger network of more nodes with smaller amounts of content per node.

Reply:
Ack. I wonder if  we could leverage DNS hints here.
We discussed having websites and gateway announce own addrs to enable clients to preconnect and skip DHT step: https://github.com/ipfs/go-ipfs/issues/6516